src/main/java/com/wavefront/sdk/grpc/WavefrontClientInterceptor.java

```diff
@@ -0,0 +1,451 @@
+    Span span = createClientSpan(methodName, method.getType().toString());
+    final ClientCallTracer tracerFactory = new ClientCallTracer(
+        Utils.getServiceName(method.getFullMethodName()), methodName,
+        shouldRecordStreamingStats(method.getType()), span);
+    ClientCall<ReqT, RespT> call =
+        next.newCall(method, callOptions.withStreamTracerFactory(tracerFactory));
+    return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(call) {
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        delegate().start(
+            new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(responseListener) {
+              @Override
+              public void onClose(Status status, Metadata trailers) {
+                tracerFactory.callEnded(status);
+                super.onClose(status, trailers);
+              }
+            },
+            headers);
+      }
+    };
+  }
+
+  @Nullable
+  private Span createClientSpan(String spanName, String methodType) {
+    if (tracer == null) {
+      return null;
+    }
+    // attempt to get active spanContext, stored in grpc context
+    Span toReturn;
+    Span activeSpan = GRPC_CONTEXT_SPAN_KEY.get();
+    if (activeSpan != null) {
+      toReturn = tracer.buildSpan(spanName).asChildOf(activeSpan.context()).start();
+    } else {
+      toReturn = tracer.buildSpan(spanName).start();
+    }
+    Tags.SPAN_KIND.set(toReturn, Tags.SPAN_KIND_CLIENT);
+    Tags.COMPONENT.set(toReturn, GRPC_CLIENT_COMPONENT);
+    toReturn.setTag(GRPC_METHOD_TYPE_KEY, methodType);
+    return toReturn;
+  }
+
+  private class ClientCallTracer extends ClientStreamTracer.Factory {
+    final String grpcService;
+    final String methodName;
+    final boolean streamingStats;
+    final Span span;
+    final AtomicBoolean streamClosed = new AtomicBoolean(false);
+    @Nullable
+    final AtomicLong requestMessageCount;
+    @Nullable
+    final AtomicLong responseMessageCount;
+    final AtomicLong requestBytes = new AtomicLong(0);
+    final AtomicLong responseBytes = new AtomicLong(0);
+    final long startTime;
+    // granular per API tags, with grpc service tag
+    final Map<String, String> allTags;
+    final Map<String, String> overallAggregatedPerSourceTags;
+    final Map<String, String> histogramAllTags;
+
+    ClientCallTracer(String grpcService, String methodName,
+                     boolean streamingStats, @Nullable Span span) {
+      this.grpcService = grpcService;
+      this.methodName = methodName;
+      this.streamingStats = streamingStats;
+      this.span = span;
+      this.startTime = System.currentTimeMillis();
+      this.requestMessageCount = streamingStats ? new AtomicLong(0) : null;
+      this.responseMessageCount = streamingStats ? new AtomicLong(0) : null;
+      // granular per RPC method metrics related tags.
+      this.allTags = ImmutableMap.<String, String>builder().
```

Comment:
ImmutableMap.Builder instance can be reused. You can first create a builder instance with cluster, service and shard and then first build the `overallAggregatedPerSourceTags`, `allTags` and then `histogramAllTags`.

Reply:
fixed, thanks!